server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -99,6 +100,10 @@ protected boolean localExecute(Request request) {
 
     protected abstract ClusterBlockException checkBlock(Request request, ClusterState state);
 
+    protected ClusterBlockException checkGlobalBlock(ClusterState clusterState) {
```

Comment:
Thinking about this some more, I think we only really care about INFEs thrown while the `STATE_NOT_RECOVERED_BLOCK` is in place, other blocks don't really matter. I think rather than introducing the `checkGlobalBlock` method (which kinda duplicates `checkBlock`) it'd be better to wrap `checkBlock` like this:

```java
    private ClusterBlockException checkBlockIfStateRecovered(Request request, ClusterState state) {
        try {
            return checkBlock(request, state);
        } catch (IndexNotFoundException e) {
            if (state.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
                // no index metadata is exposed yet, but checkBlock depends on an index, so keep trying until the cluster forms
                assert GatewayService.STATE_NOT_RECOVERED_BLOCK.contains(ClusterBlockLevel.METADATA_READ);
                assert state.blocks().global().stream().allMatch(ClusterBlock::retryable);
                return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
            } else {
                throw e;
            }
        }
    }
```